wemake_python_styleguide/logic/tree/classes.py

```diff
@@ -34,3 +39,87 @@ def is_forbidden_super_class(class_name: Optional[str]) -> bool:
     if class_name in ALLOWED_BUILTIN_CLASSES:
         return False
     return is_builtin_name(class_name)
+
+
+def get_attributes(
+    node: ast.ClassDef,
+) -> Tuple[List[types.AnyAssign], List[ast.Attribute]]:
+    """Returns all non annotated class and instance attributes of a class."""
+    return get_class_attributes(node), get_instance_attributes(node)
+
+
+def get_all_attributes_stripped(node: ast.ClassDef) -> Set[str]:
+    """
+    Returns names all class and instance attributes of a class stripped.
+
+    Removes all leading underscores from attribute names.
+
+    """
+    class_attributes = get_annotated_class_attributes(node)
+    instance_attributes = get_instance_attributes(node)
+    return {
+        class_attribute.lstrip('_') for class_attribute
+        in name_nodes.flat_variable_names(class_attributes)
+    }.union({
+        instance.attr.lstrip('_') for instance
+        in instance_attributes
+    })
+
+
+def get_class_attributes(
+    node: ast.ClassDef,
+) -> List[Union[ast.Assign, ast.AnnAssign]]:
+    """Returns all non annotated class attributes of a class."""
+    class_attributes = []
+    for sub in ast.walk(node):
+        is_correct_context = nodes.get_context(sub) == node
+        has_value = getattr(sub, 'value', None)
+        if isinstance(sub, AssignNodes) and has_value and is_correct_context:
+            class_attributes.append(sub)
+
+    return class_attributes
+
+
+def get_annotated_class_attributes(
+    node: ast.ClassDef,
+) -> List[Union[ast.Assign, ast.AnnAssign]]:
+    """Returns all class attributes of a class."""
+    class_attributes = []
+    for sub in ast.walk(node):
+        is_correct_context = nodes.get_context(sub) == node
+        has_value = getattr(sub, 'value', None)
+        if isinstance(sub, AssignNodes) and has_value and is_correct_context:
+            class_attributes.append(sub)
+            continue
+        if isinstance(sub, ast.AnnAssign) and is_correct_context:
+            class_attributes.append(sub)
+
+    return class_attributes
+
+
+def get_instance_attributes(node: ast.ClassDef) -> List[ast.Attribute]:
+    """Returns all instance attributes of a class."""
+    instance_attributes = []
+    for sub in ast.walk(node):
+        if isinstance(sub, ast.Attribute) and isinstance(sub.ctx, ast.Store):
```

Comment:
This is missing `'self'` part check. It will also count `other.attr` as instance attribute.

Reply:
Should it matter for this violation? The naming should still be disallowed.

Reply:
Yes, because this code should not produce any violations:

```python
class Some(object):
    def __init__(self, other):
        other.attr = self.some()

   def get_attr(self):
        return something.unrelated()
```

Here `get_attr` technically is not a getter. But this check will fail. See https://github.com/wemake-services/wemake-python-styleguide/blob/master/wemake_python_styleguide/logic/tree/classes.py#L44

Reply:
If they are, can you give some example / test cases which i can use?

Reply:
Guess i could just use the new implementation.

Reply:
Shouldn't the shadowing violation also apply for AnnAssign type attributes?

Reply:
Nope, shadowing should not report `AnnAssign`:

```python
class Test(object):
    attr: int   # class does not have `attr` in `__dict__` at this point

    def __init__(self):
         self.attr = 1  # now it is in `__dict__` of this instance
```

Reply:
But! If `AnnAssign` has a value. Than - yes! That's a bug.

Can you please report it?

Reply:
sure!

Reply:
Hi this seems to have already been checked:
´´´
class_attribute_annotated = """
class ClassWithAttrs(object):
    {0}: int = 0

    def __init__(self) -> None:
        self.{1} = 2
"""
´´´´
This test case raises the shadowing violation as it should. Seems to be working without checking for AnnAssign.

Reply:
Yes, it is covered indeed 👍